source/dub/generators/generator.d

```diff
@@ -332,23 +332,24 @@ ProjectGenerator createProjectGenerator(string generator_type, Project project)
 /**
 	Runs pre-build commands and performs other required setup before project files are generated.
 */
-private void prepareGeneration(in Package pack, in GeneratorSettings settings, in BuildSettings buildsettings)
+private void prepareGeneration(in Package pack, in Project proj, in GeneratorSettings settings,
+	in BuildSettings buildsettings)
 {
-	if (buildsettings.preGenerateCommands.length) {
+	if (buildsettings.preGenerateCommands.length && !PackagesUsed.has(pack.name)) {
 		logInfo("Running pre-generate commands for %s...", pack.name);
-		runBuildCommands(buildsettings.preGenerateCommands, pack, settings, buildsettings);
+		runBuildCommands(buildsettings.preGenerateCommands, pack, proj, settings, buildsettings);
 	}
 }
 
 /**
 	Runs post-build commands and copies required files to the binary directory.
 */
-private void finalizeGeneration(in Package pack, in GeneratorSettings settings, in BuildSettings buildsettings,
-	Path target_path, bool generate_binary)
+private void finalizeGeneration(in Package pack, in Project proj, in GeneratorSettings settings,
+	in BuildSettings buildsettings, Path target_path, bool generate_binary)
 {
-	if (buildsettings.postGenerateCommands.length) {
+	if (buildsettings.postGenerateCommands.length && !PackagesUsed.has(pack.name)) {
```

Comment:
The lack of abstraction seems very dirty there. Basic encapsulation principles apply: Any future changes that might be needed to the way the info is stored/passed would be messy and error-prone, plus it's mixing high-level intent with low-level implementation in the same line of code.